jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java

```diff
@@ -125,7 +131,48 @@
    */
   @Nullable
   T call() throws IOException, RegistryException {
-    return call(initialRequestUrl);
+    return callWithAllowInsecureRegistryHandling(initialRequestUrl);
+  }
+
+  @Nullable
+  private T callWithAllowInsecureRegistryHandling(URL url) throws IOException, RegistryException {
+    if (allowInsecureRegistries) {
+      return possbilyInsecureCall(url);
+    }
+
+    try {
+      if (!isHttpsProtocol(url)) {
+        throw new InsecureRegistryException(url);
+      }
+      return call(url, connectionFactory);
+    } catch (SSLPeerUnverifiedException ex) {
+      throw new InsecureRegistryException(url);
+    }
+  }
+
+  @Nullable
+  private T possbilyInsecureCall(URL url) throws IOException, RegistryException {
+    Preconditions.checkState(allowInsecureRegistries);
+    try {
+      return call(url, connectionFactory);
+
+    } catch (SSLPeerUnverifiedException exception) {
+      try {
+        if (insecureConnectionFactory == null) {
+          insecureConnectionFactory = Connection.getInsecureConnectionFactory();
```

Comment:
Hmm, it seems that `insecureConnectionFactory` doesn't need to be an instance variable.

Reply:
In fact, `connectionFactory` doesn't have to be so either, but it was there from the beginning, and I believe it is for mocking `Connection`. `insecureConnectionFactory` is there for the same reason.

But I guess you will wonder why the code has this null check.

```java
        Connection.getConnectionFactory(),
        null /* might never be used, so create lazily to delay throwing potential GeneralSecurityException */);
```

So this is to avoid the potential `GeneralSecurityException` when creating the `insecureConnectionFactory`. For example, if the user never sets `allowInsecureRegistries` or the server was secure even if it was set, we can proceed without an error in the situation where `GeneralSecurityException` will otherwise be thrown if blindly created at the beginning.